### DISCUSSION TITLE
can I set the timestamp?

### DISCUSSION THREAD

**Discussion by joyously** [jj-vcs/jj#1387] [General]
Once upon a time, I saved several versions of some code in zip files, before I knew about version control. Later I was able to create the history of the changes in Bazaar, using the `--commit-time` flag of its `commit` command. Looking at Git, I see its `commit` has a `--date` flag.

Does `jj` have a way to explicitly set the timestamp? or author?

**Comment by martinvonz** [comment]
No, that's not supported yet. We have #328 for that. It's easy to implement; the hard part is to decide what the UX should be.

**Comment by programandala-net** [comment]
I'm a Mercurial user and I'm trying Jujutsu.  It seems cleverly designed, and it looks really promising.

During my tests I've had the same need @joyously describes.

I have read [#328](https://github.com/martinvonz/jj/issues/328) and searched the [command-line documentation](https://martinvonz.github.io/jj/latest/cli-reference/) for a way to change the timestamp of a change, and I was surprised that `jj describe` doesn't have a `--date` option yet, but only `--reset-author`, which resets the timestamp to the current time.

**Reply by martinvonz** [reply]
There's an undocumented way: `JJ_TIMESTAMP=2024-10-14T13:27:32-08:00 jj describe --reset-author`.

I think it's just very rare that people need to set the timestamps on a commit. What's your use case?

**Reply by ilyagr** [reply]
In bash, you can also try `JJ_TIMESTAMP=$(date -Iseconds)` for the current time. It can also convert time from other formats.

**Reply by ilyagr** [reply]
But also, adding a `--reset-date` option to `jj describe` makes a lot of sense to me. It's just that nobody has done it yet. In other words, I don't know of any concerns that people would have about it, though I don't know everybody's mind.

**Reply by programandala-net** [reply]
> There's an undocumented way: `JJ_TIMESTAMP=2024-10-14T13:27:32-08:00 jj describe --reset-author`.

Thank you. That solves the problem.

I had tried `JJ_DATE` and `JJ_TIMESTAMP` blindly, by analogy with `JJ_AUTHOR`, but it didn't work.  Now I see the reason is I ommited some parts (seconds and time zone, or even hours and minutes) supposing the missing precision would be completed with default values, which is the behaviour of Mercurial's `--date` and, if I recall clearly, also Fossil's `--date-override` and Git's `--date` (I used both of them some years ago).

> I think it's just very rare that people need to set the timestamps on a commit. What's your use case?

I agree, but `--date` is necessary to build repositories out of backups of old projects that were developed without a VCS, or to start a new repository with ancient stuff, thus preserving the history and original dates.  I've done it many times, with the help of some scripts, to build Git, Fossil and Mercurial repos. Also, sometimes I used this option in order to fix time zones, especially when converting a repo from one VCS to other.

Anyway, using an environment variable and `--reset-author` is fine, but the usual specific option would be simpler and clearer:

```
JJ_TIMESTAMP=2024-10-14T13:27:32-08:00 jj describe --reset-author
jj describe --date 2024-10-14T13:27:32-08:00 
```

In fact, given the clear design of Jujutsu's command line (with its great combination of the good features of Mercurial and Fossil), the same criteria would be applicable to the rest of the `describe` options:

```
--reset — Reset the author and email to the configured user; reset date to the current time.
--reset-author — Reset the author [and email?] to the configured user.
--reset-date — Reset the timestamp to the current time.
--author <AUTHOR> — Set author [and email?] to the provided string.
--date <DATE> — Set date to the provided string.
```

Probably it makes sense that `--reset-author` and `--author` change also the email, since the need to change only one of them is improbable; but given Jujustu, contrary to Mercurial, keeps author and email apart, maybe reserving `--author` and `--reset-author` for the author and adding the following options for the email would make sense as well:

```
--reset-email — Reset the email to the configured user.
--email <EMAIL> — Set email to the provided string.
```

**Reply by programandala-net** [reply]
> There's an undocumented way: `JJ_TIMESTAMP=2024-10-14T13:27:32-08:00 jj describe --reset-author`.

I just discovered that method changes also the timestamps of the descendants.  I guess the value of the environment variable is used when they are automatically rebased.

**Reply by martinvonz** [reply]
Yes, that's correct. Makes sense to add a flag specifically for setting the timestamp, or maybe even two separate `--author-date` and `--committer-date`.

**Comment by andyleiserson** [comment]
At least some of these changes are now possible with [`jj metaedit`](https://jj-vcs.github.io/jj/latest/cli-reference/#jj-metaedit).
